lnwallet/wallet.go

```diff
@@ -431,6 +431,56 @@ func (l *LightningWallet) Shutdown() error {
 	return nil
 }
 
+// PublishTransaction wraps the wallet controller tx publish method with an
+// extra rebroadcaster layer if the sub-system is configured.
+func (b *LightningWallet) PublishTransaction(tx *wire.MsgTx, label string) error {
+	err := b.WalletController.PublishTransaction(tx, label)
+	if err != nil {
+		return err
+	}
+
+	// If we don't have rebroadcaster then we can exit early.
+	if b.Cfg.Rebroadcaster == nil {
+		return nil
+	}
+
+	// TODO(roasbeef): want diff height actually? no context though
+	_, bestHeight, err := b.Cfg.ChainIO.GetBestBlock()
+	if err != nil {
+		return err
+	}
+
+	// We'll launch a goroutine to add the transaction to the
+	// rebroadcaster, then wait until its confirmed to signal that it no
+	// longer needs to be rebroadcast.
+	if err := b.Cfg.Rebroadcaster.Broadcast(tx); err != nil {
+		return err
+	}
+
+	txHash := tx.TxHash()
+	go func() {
+		txConf, err := b.Cfg.Notifier.RegisterConfirmationsNtfn(
+			&txHash, tx.TxOut[0].PkScript, 1, uint32(bestHeight),
+		)
+		if err != nil {
+			return
+		}
+
+		select {
+		case <-txConf.Confirmed:
+			// TODO(roasbeef): also want to remove from
+			// rebroadcaster if conflict happens...deeper wallet
+			// integration?
+			b.Cfg.Rebroadcaster.MarkAsConfirmed(tx.TxHash())
```

Comment:
This could also be nested inside `Rebroadcaster` if `Broadcast` takes care of the confirmation ntfn.

Reply:
Perhaps never expose `MarkAsConfirmed` and take care of it internally?

Reply:
Things were crafted like this on the neutrino side as the core p2p logic is decoupled from the notion of a "rescan" or a wallet. This decoupling led to this design where the caller is responsible for telling the broadcaster to stop. It can grow to also fetch blocks and see if things have been confirmed or not, but then that'll likely end up with a lot of duplicate block fetching as the main rescan loop does that.